x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/fielddata/CartesianShapeValues.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.spatial.index.fielddata;
+
+import org.apache.lucene.geo.Component2D;
+import org.apache.lucene.geo.XYGeometry;
+import org.apache.lucene.geo.XYPoint;
+import org.elasticsearch.search.aggregations.support.ValuesSourceType;
+import org.elasticsearch.xpack.spatial.common.CartesianPoint;
+import org.elasticsearch.xpack.spatial.index.mapper.CartesianShapeIndexer;
+import org.elasticsearch.xpack.spatial.search.aggregations.support.CartesianShapeValuesSourceType;
+
+import java.io.IOException;
+
+/**
+ * A stateful lightweight per document geo values.
+ * To iterate over values in a document use the following pattern:
+ * <pre>
+ *   MultiGeoValues values = ..;
+ *   // for each docID
+ *   if (values.advanceExact(docId)) {
+ *     GeoValue value = values.value()
+ *     final int numValues = values.count();
+ *     // process value
+ *   }
+ * </pre>
+ *
+ * There is just one value for one document.
+ */
+public abstract class CartesianShapeValues extends ShapeValues {
+
+    public static CartesianShapeValues EMPTY = new CartesianShapeValues() {
+        private final CartesianShapeValuesSourceType DEFAULT_VALUES_SOURCE_TYPE = CartesianShapeValuesSourceType.instance();
+
+        @Override
+        public boolean advanceExact(int doc) {
+            return false;
+        }
+
+        @Override
+        public ValuesSourceType valuesSourceType() {
+            return DEFAULT_VALUES_SOURCE_TYPE;
+        }
+
+        @Override
+        public CartesianShapeValue value() {
+            throw new UnsupportedOperationException();
+        }
+    };
+
+    /**
+     * Creates a new {@link CartesianShapeValues} instance
+     */
+    protected CartesianShapeValues() {
+        super(CoordinateEncoder.CARTESIAN, CartesianShapeValues.CartesianShapeValue::new, new CartesianShapeIndexer("missing"));
+    }
+
+    /** thin wrapper around a {@link GeometryDocValueReader} which encodes / decodes values using
+     * the Geo decoder */
+    public static class CartesianShapeValue extends ShapeValues.ShapeValue {
```

Comment:
There is something that is not right here because ShapeValues.ShapeValue contains geo specific code. For example this means that CartesianShapeValue supports calling #relate(LatLonGeometry) which is not right. I think we ought to move the geo specific methods out of ShapeValue and only leave the generic ones?

Reply:
There is no geo-specific code in `ShapeValues` or in `ShapeValues.ShapeValue`. I double checked and the only mistake is in class naming. For some reason classes are named `LatLon*` when they have no geo-specific behaviour. I commented on this above, and also pushed a new commit that further reduces the code duplication in `ShapeValues`.

If you can find any geo-specific method in `ShapeValues` or in `ShapeValues.ShapeValue`, please let me know, because I cannot find them. It was made generic in the previous PR, there is nothing geo-specific left.